### ISSUE TITLE
Use of SWAT in Jupyter

### ISSUE THREAD

**Issue by fabricepollet** [sassoftware/python-swat#9]
Hello
Do you ever experience problem with the use of SWAT in jupyter?

I use the following code

# Import necessary packages and modules
import swat
hostname = 'localhost'
port=5570
username='maou'
password='XXXXX'
# Set the connection by specifying the hostname, port, username, and password
conn = swat.CAS(hostname,port,username,password)

And I receive the following

JSONDecodeError                           Traceback (most recent call last)
/opt/anaconda3/lib/python3.6/site-packages/swat/cas/rest/connection.py in __init__(self, hostname, port, username, password, soptions, error)
    225                                          'cas/sessions'), data=b'')
--> 226                 out = json.loads(a2u(res.text, 'utf-8'))
    227                 if res.status_code != 200 and 'error' in out:

/opt/anaconda3/lib/python3.6/json/__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    353             parse_constant is None and object_pairs_hook is None and not kw):
--> 354         return _default_decoder.decode(s)
    355     if cls is None:

/opt/anaconda3/lib/python3.6/json/decoder.py in decode(self, s, _w)
    338         """
--> 339         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    340         end = _w(s, end).end()

/opt/anaconda3/lib/python3.6/json/decoder.py in raw_decode(self, s, idx)
    356         except StopIteration as err:
--> 357             raise JSONDecodeError("Expecting value", s, err.value) from None
    358         return obj, end

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

SWATError                                 Traceback (most recent call last)
<ipython-input-3-11c23c7c568d> in <module>()
      6 password='Viya123'
      7 # Set the connection by specifying the hostname, port, username, and password
----> 8 conn = swat.CAS(hostname,port,username,password)

/opt/anaconda3/lib/python3.6/site-packages/swat/cas/connection.py in __init__(self, hostname, port, username, password, session, locale, nworkers, name, authinfo, protocol, **kwargs)
    274                           a2n(soptions), self._sw_error)
    275                 if protocol in ['http', 'https']:
--> 276                     self._sw_connection = rest.REST_CASConnection(*params)
    277                 else:
    278                     self._sw_connection = clib.SW_CASConnection(*params)

/opt/anaconda3/lib/python3.6/site-packages/swat/cas/rest/connection.py in __init__(self, hostname, port, username, password, soptions, error)
    237                     self._results.clear()
    238         except Exception as exc:
--> 239             raise SWATError(str(exc))
    240 
    241     def invoke(self, action_name, kwargs):

SWATError: Expecting value: line 1 column 1 (char 0)

thanks for your help
Fabrice

**Comment by kesmit13** [comment]
When a new connection is made to a CAS server, it attempts to determine the server protocol (binary or REST).  When connecting to a binary port (as you are in this example), it should fail when it tests for an HTTP connection to that port.  For some reason, that isn't happening and it's trying to interpret the binary results.  You can force a binary connection by specifying "protocol='cas'" on your CAS connection.  Try that and see what happens.

**Comment by fabricepollet** [comment]
Thanks Kevin.
It works well.

De : Kevin D Smith [mailto:notifications@github.com]
Envoyé : jeudi 10 août 2017 18:54
À : sassoftware/python-swat
Cc : Pollet, Fabrice; Author
Objet : Re: [sassoftware/python-swat] Use of SWAT in Jupyter (#9)


When a new connection is made to a CAS server, it attempts to determine the server protocol (binary or REST). When connecting to a binary port (as you are in this example), it should fail when it tests for an HTTP connection to that port. For some reason, that isn't happening and it's trying to interpret the binary results. You can force a binary connection by specifying "protocol='cas'" on your CAS connection. Try that and see what happens.

**Comment by kesmit13** [comment]
I'll still look into why the auto-detection of the protocol didn't work.

**Comment by kesmit13** [comment]
What version of SWAT are you running?  I still can't seem to reproduce this problem.

```
import swat
print(swat.__version__)
```

**Comment by kesmit13** [comment]
If you could send me the traceback for the following code, that would be helpful too. 

```
from six.moves import urllib
req = urllib.request.Request('http://localhost:5570/cas')
with urllib.request.urlopen(req) as res:
    pass
```

**Comment by fabricepollet** [comment]
Hello

Here is the result :
1.2.0



De : Kevin D Smith [mailto:notifications@github.com]
Envoyé : mercredi 16 août 2017 20:33
À : sassoftware/python-swat
Cc : Pollet, Fabrice; Author
Objet : Re: [sassoftware/python-swat] Use of SWAT in Jupyter (#9)


What version of SWAT are you running? I still can't seem to reproduce this problem.

import swat

print(swat.__version__)

**Comment by fabricepollet** [comment]
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-2-90e836262682> in <module>()
      1 from six.moves import urllib
      2 req = urllib.request.Request('http://localhost:5570/cas')
----> 3 with urllib.request.urlopen(req) as res:
      4     pass

/opt/anaconda3/lib/python3.6/urllib/request.py in urlopen(url, data, timeout, cafile, capath, cadefault, context)
    221     else:
    222         opener = _opener
--> 223     return opener.open(url, data, timeout)
    224
    225 def install_opener(opener):

/opt/anaconda3/lib/python3.6/urllib/request.py in open(self, fullurl, data, timeout)
    530         for processor in self.process_response.get(protocol, []):
    531             meth = getattr(processor, meth_name)
--> 532             response = meth(req, response)
    533
    534         return response

/opt/anaconda3/lib/python3.6/urllib/request.py in http_response(self, request, response)
    640         if not (200 <= code < 300):
    641             response = self.parent.error(
--> 642                 'http', request, response, code, msg, hdrs)
    643
    644         return response

/opt/anaconda3/lib/python3.6/urllib/request.py in error(self, proto, *args)
    568         if http_err:
    569             args = (dict, 'default', 'http_error_default') + orig_args
--> 570             return self._call_chain(*args)
    571
    572 # XXX probably also want an abstract factory that knows when it makes

/opt/anaconda3/lib/python3.6/urllib/request.py in _call_chain(self, chain, kind, meth_name, *args)
    502         for handler in handlers:
    503             func = getattr(handler, meth_name)
--> 504             result = func(*args)
    505             if result is not None:
    506                 return result

/opt/anaconda3/lib/python3.6/urllib/request.py in http_error_default(self, req, fp, code, msg, hdrs)
    648 class HTTPDefaultErrorHandler(BaseHandler):
    649     def http_error_default(self, req, fp, code, msg, hdrs):
--> 650         raise HTTPError(req.full_url, code, msg, hdrs, fp)
    651
    652 class HTTPRedirectHandler(BaseHandler):

HTTPError: HTTP Error 403: Forbidden


De : Kevin D Smith [mailto:notifications@github.com]
Envoyé : mercredi 16 août 2017 20:46
À : sassoftware/python-swat
Cc : Pollet, Fabrice; Author
Objet : Re: [sassoftware/python-swat] Use of SWAT in Jupyter (#9)


If you could send me the traceback for the following code, that would be helpful too.

from six.moves import urllib

req = urllib.request.Request('http://localhost:5570/cas')

with urllib.request.urlopen(req) as res:

    pass

**Comment by kesmit13** [comment]
That's interesting.  I would have expected a ConnectionResetError from the binary port rather than an HTTPError.  Unfortunately, the fact that it raises an HTTPError is the reason that it thinks it's an HTTP connection.  I still can't reproduce it on my end, but I'm still looking into it.

**Comment by kesmit13** [comment]
While I couldn't replicate your issue, we changed the way that protocol detection is done in v1.2.1.  You might want to try it and see if this fixes it for you.

**Comment by fabricepollet** [comment]
Thanks Kevin.
It works perfectly with v1.2.1.

Fabrice

De : Kevin D Smith [mailto:notifications@github.com]
Envoyé : mercredi 13 septembre 2017 23:08
À : sassoftware/python-swat
Cc : Pollet, Fabrice; Author
Objet : Re: [sassoftware/python-swat] Use of SWAT in Jupyter (#9)


While I couldn't replicate your issue, we changed the way that protocol detection is done in v1.2.1. You might want to try it and see if this fixes it for you.

**Comment by kesmit13** [comment]
Fixed in v1.2.1.
